### PR TITLE
[CONSVC-1603]: add feature flags as tags to metrics

### PR DESCRIPTION
**Pull request by hackebrot**
## Description

This pull request integrates feature flags with StatsD metrics. 📈

Feature flag decisions are now automatically recorded and added as tags to emitted metrics.

## Issue(s)

[CONSVC-1603](https://mozilla-hub.atlassian.net/browse/CONSVC-1603)

### CODE REVIEW

File: merino/metrics.py [hunk 1]
```diff
@@ -1,19 +1,123 @@
...
 
 logger = logging.getLogger(__name__)
 
+# Type definitions for tags in aiodogstatsd metrics
+MTagKey = str
+MTagValue = Union[float, int, str]
+MTags = Mapping[MTagKey, MTagValue]
+
+# The following methods are added to the `Client` class by the meta class
+SUPPORTED_METHODS: list[str] = ["increment", "timing", "timeit_task"]
```
Comment (quiiver) [inline merino/metrics.py:21]: is there a reason we wouldn't want to add `decrement, gauge, histogram, and distribution` to this list?

File: merino/web/api_v1.py [hunk 2]
```diff
@@ -54,6 +53,24 @@ async def suggest(
...
+        if feature_flags.is_enabled("test-perc-enabled"):
+            # TODO: Do we expect a decision for a `random` scheme flag to be
+            # consistent across the handling of a request? Do we expect to call
+            # is_enabled() multiple times for a feature flag or just once have
+            # retrieve the recorded decision for future calls in
+            # @record_decision?
+            print("test-perc-enabled is enabled")
+
+        if feature_flags.is_enabled("test-perc-enabled-session"):
+            print("test-perc-enabled-session is enabled")
```
Comment (quiiver) [inline merino/web/api_v1.py:72]: is this for debugging?

File: merino/metrics.py (hunk 1 above)
Comment (hackebrot) [reply merino/metrics.py]: Not at all! I've only added the ones we were currently using.

Just pushed a commit to add the remaining metric methods.

File: merino/web/api_v1.py (hunk 2 above)
Comment (hackebrot) [reply merino/web/api_v1.py:72]: Yes, I've used this for debugging while working on these changes. I left it in for demonstration purposes.

I'll make sure to remove this before marking this as ready for review. 👍🏻

File: merino/metrics.py [hunk 3]
```diff
@@ -1,19 +1,137 @@
...
+    kwargs["tags"] = {**feature_flags_tags, **tags}
+
+    return wrapped_method(*args, **kwargs)
+
+
+class ClientMeta(type):
+    """Metaclass that decorates Client methods with add_feature_flags."""
+
+    def __new__(
+        cls: type[_M],
```
Comment (Trinaa) [inline merino/metrics.py:73]: My IDE indicates a warning for this line:
image

I'm not familiar with metaclass convention or if this warning is applicable?

File: merino/metrics.py [hunk 4]
```diff
@@ -1,19 +1,137 @@
...
+):
+    """Add feature flag decisions as tags when recording metrics."""
+    # Tags added manually to the metrics client call
+    tags = kwargs.pop("tags", {})
+
+    # Generate tags based on the recorded feature flag decisions
+    feature_flags_tags = feature_flags_as_tags(instance.feature_flags)
+
+    # The order is important here. Feature flag tags added manually take
+    # precendence over the auto-generated ones.
```
Comment (Trinaa) [inline merino/metrics.py:63]:
```suggestion
    # The order is important here. Feature flag tags added manually take
    # precedence over the auto-generated ones.
```

File: merino/metrics.py [hunk 5]
```diff
@@ -1,19 +1,137 @@
...
+        bases: tuple[type, ...],
+        namespace: dict[str, Any],
+        **kwargs: Any,
+    ) -> _M:
+        """Create a new class with decorated methods."""
+
+        def method_proxy(method_name: str) -> Callable:
+            """Return a method that proxies to correct method of the StatsD client."""
+
+            def client_method(instance: "Client", *args, **kwargs) -> Any:
```
Comment (Trinaa) [inline merino/metrics.py:84]: The `method_name` and `kwargs` arguments are used in the outer-scope `__new__` method as well. Should they be named differently? Or will this break convention?

File: tests/unit/test_metrics.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+@pytest.fixture(name="client")
+def fixture_client(statsd_mock):
+    """Return a metrics client."""
+    return Client(
+        statsd_client=statsd_mock,
+        feature_flags=FeatureFlags(),
+    )
+
+
+def test_unsupported_method(client, statsd_mock):
```
Comment (Trinaa) [inline tests/unit/test_metrics.py:25]: Is there value in typing these parameters or maybe renaming `client` to `metric_client`? 

I'm starting to observe how many clients we have in this project and wondering if we should start adding more context.

**Review by Trinaa — COMMENTED** [review]
A few nits and questions for you.

I will also note that I had to upgrade setuptools in order to execute the tests locally on this branch.

File: merino/metrics.py (hunk 3 above)
Comment (pjenvey) [reply merino/metrics.py:73]: `cls` is more common (historically)

Comment (hackebrot) [reply merino/metrics.py:73]: It's common to name the first positional argument of `__new__` as `cls`:
https://docs.python.org/3/reference/datamodel.html?highlight=__new__#object.__new__

I assume PyCharm suggests `mcs` because the class here is a metaclass. However I can't seem to find any uses of that name in the Python documentation or the CPython source code:
https://github.com/python/cpython/search?q=mcs

I'll rename it to `mcls` to indicate that the object is in fact an instance of the metaclass.

File: tests/unit/test_metrics.py (hunk 6 above)
Comment (hackebrot) [reply tests/unit/test_metrics.py:25]: Yes, you are right. That's a good observation! I'll update this. 👍🏻

File: merino/metrics.py (hunk 5 above)
Comment (hackebrot) [reply merino/metrics.py]: I have renamed the arguments. Good idea!

File: merino/metrics.py [hunk 7]
```diff
@@ -1,19 +1,150 @@
...
+    """Return a representation of feature flags decisions."""
+    return {
+        f"{FLAGS_PREFIX}.{name}": int(decision)
+        for name, decision in feature_flags.decisions.items()
+    }
+
+
+@decorator
+def add_feature_flags(
+    wrapped_method: Callable, instance: "Client", args: tuple, kwargs: dict
```
Comment (quiiver) [inline merino/metrics.py:49]: where does the `instance` arg come from? `wrapt`?

Comment (quiiver) [reply merino/metrics.py:49]: Also would it be possible to use the TypeVar `C` here?

Comment (hackebrot) [reply merino/metrics.py:49]: Yes, that comes from wrapt, see:
https://wrapt.readthedocs.io/en/latest/decorators.html#creating-decorators

**Comment by ncloudioj** [comment]
@hackebrot Sorry for the delayed response.

I've skimmed over the PR, and it looks good. I will take another closer look just in a bit.

Also, I've run some preliminary benchmarks locally, and have seen some performance improvements 🚀 so far:

Running 100VU for 60 seconds.

The main branch: 2,531 RPS, 52.18ms latency (P95)
This branch: 2,858 RPS, 47.2ms

Will do it with a few more realistic scenarios to verify the results.

**Comment by ncloudioj** [comment]
Okay, more benchmark results:

Tl;dr: local benchmarks show that this branch outperforms (10% relative uplift w.r.t throughput & latency) the main branch in both testing scenarios (w/ & w/o feature flags).

## Without feature flags
### Main branch
- RPS: 2,531
- Latency (P95): 52.18ms

### This branch
- RPS: 2,858
- Latency (P95): 47.20ms

## With three feature flags
All results below are using the following feature flag configs:
```
[development.flags.feature-a]
enabled = 1
scheme = 'session'

[development.flags.feature-b]
enabled = 0.5
scheme = 'session'

[development.flags.feature-c]
enabled = 0.5
scheme = 'session'
```

### Main branch with 3 feature flags
 - RPS: 2,504
- Latency (P95): 58.00ms

### This branch with 3 feature flags
 - RPS: 2,814
- Latency (P95): 53.00ms

File: merino/metrics.py [hunk 8]
```diff
@@ -1,19 +1,149 @@
...
 
 from merino.config import settings
+from merino.featureflags import FeatureFlags
 
 logger = logging.getLogger(__name__)
 
+# Type definitions for tags in aiodogstatsd metrics
+MetricTagKey = str
+MetricTagValue = Union[float, int, str]
+MetricTags = Mapping[MetricTagKey, MetricTagValue]
```
Comment (ncloudioj) [inline merino/metrics.py:18]: Nit: looks like `MetricTagKey` and `MetricTagValue` are only referenced once by MetricTags. Do we really need those type aliases here? I agree that it looks slightly clearer than `MetricsTags = Mapping[str, float | int | str]`, but all those aliases also make this module quite hard to follow, at least that was the case when I first looked at this code. I was kinda overwhelmed by those type declarations, although the actual code below was quite easy to understand.

Shall we consider simplifying those?

File: merino/metrics.py [hunk 9]
```diff
@@ -1,19 +1,149 @@
...
+    kwargs["tags"] = {**feature_flags_tags, **tags}
+
+    return wrapped_method(*args, **kwargs)
+
+
+class ClientMeta(type):
+    """Metaclass that decorates Client methods with add_feature_flags."""
+
+    def __new__(
+        mcls: type[M],
```
Comment (ncloudioj) [inline merino/metrics.py:81]: Certain linters might issue a warning for not using `cls` as the first argument for `__new__`, any reason for not using `cls`?

File: merino/web/api_v1.py [hunk 10]
```diff
@@ -54,6 +51,14 @@ async def suggest(
     Returns:
     A list of suggestions or an empty list if nothing was found.
     """
+    # Do you plan to release code behind a feature flag? Uncomment the following
+    # line to get access to feature flags and then check if your feature flag is
+    # enabled for this request by calling feature_flags.is_enabled("example").
+    # Then unskip test_feature_flags and update it with your feature flag.
```
Comment (ncloudioj) [inline merino/web/api_v1.py:57]: Hmm, what do you mean by `unskip test_feature_flags`?

File: merino/web/api_v1.py [hunk 11]
```diff
@@ -54,6 +51,14 @@ async def suggest(
     Returns:
     A list of suggestions or an empty list if nothing was found.
     """
+    # Do you plan to release code behind a feature flag? Uncomment the following
```
Comment (ncloudioj) [inline merino/web/api_v1.py:54]: I find this note super useful, but could you move (or copy) it to a dedicated document under the docs directory? Then we can use that doc to describe how to use feature flags and other related details in this repo. Thanks in advance!

**Review by ncloudioj — APPROVED** [review]
Looks good, r+ wc.

File: merino/metrics.py (hunk 9 above)
Comment (hackebrot) [reply merino/metrics.py:81]: See above conversation about this: https://github.com/mozilla-services/merino-py/pull/92#discussion_r1011464007

File: merino/web/api_v1.py (hunk 10 above)
Comment (hackebrot) [reply merino/web/api_v1.py:57]: I have updated the comment to the following. Let me know if this is better! 😄 

```python
# Do you plan to release code behind a feature flag? Uncomment the following
# line to get access to feature flags and then check if your feature flag is
# enabled for this request by calling feature_flags.is_enabled("example").
# feature_flags: FeatureFlags = request.scope[ScopeKey.FEATURE_FLAGS]

# Then remove the `pytest.mark.skip` decorator from the test
# `test_feature_flags` and update it with your feature flag.
```

File: merino/metrics.py (hunk 8 above)
Comment (hackebrot) [reply merino/metrics.py]: I don't have a strong opinion about this. I pushed a commit that deletes `MetricTagKey` and `MetricTagValue`.

File: merino/metrics.py [hunk 12]
```diff
@@ -29,7 +157,7 @@ async def configure_metrics() -> None:
     await client.connect()
 
 
-class _LocalDatagramLogger(DatagramProtocol):
+class _LocalDatagramLogger(aiodogstatsd.client.DatagramProtocol):
```
[the remark this answers is not in the document]
Comment (hackebrot) [reply merino/metrics.py:160]: Yes! I created https://mozilla-hub.atlassian.net/browse/CONSVC-2089 for that. ✅

File: tests/unit/web/test_metrics.py [hunk 13]
```diff
@@ -37,34 +40,88 @@ def corrupt_provider():
...
+    """Test that feature flags are added for successful requests."""
+    report = mocker.patch.object(aiodogstatsd.Client, "_report")
+
+    client.get(url)
+
+    want = {metric_key: tags for metric_key in metric_keys}
+
+    # TODO: This is not great. We're relying on internal details of aiodogstatsd
+    # here Can we record calls to the metrics client instead?
+    got = {call.args[0]: [*call.args[3].keys()] for call in report.call_args_list}
```
[the remark this answers is not in the document]
Comment (hackebrot) [reply tests/unit/web/test_metrics.py:106]: Created https://mozilla-hub.atlassian.net/browse/CONSVC-2090 ✅

**Comment by hackebrot** [comment]
Thanks for the reviews, team! 📦🚀
